core/src/main/java/hudson/TcpSlaveAgentListener.java

```diff
@@ -116,6 +123,16 @@ public void run() {
     public void shutdown() {
         shuttingDown = true;
         try {
+            SocketAddress localAddress = serverSocket.getLocalAddress();
```

Comment:
Hmm the  ServerSocketChannel should really be in [non blocking mode](https://docs.oracle.com/javase/7/docs/api/java/nio/channels/spi/AbstractSelectableChannel.html#configureBlocking(boolean)) and should check for a null return and wait appropriately (and then all you ned here is a notify).